### ISSUE TITLE
Question about normalisation and unbinding

### ISSUE THREAD

**Issue by danmichaeljones** [hyperdimensional-computing/torchhd#183]
Hey, thanks for making this package available, it's been really useful to me as someone whose only just come across VSAs! I had a question about normalisation and unbinding that I'm hoping is just down to my misunderstanding.

I have been testing out a workflow of binding together many pairs of key / value vectors, bundling them together into a single vector, and then querying back one of the keys by unbinding. Using `FHRRTensor`, I'm doing this as follows:

```
import torch
import torchhd as thd
import numpy as np

N_VALUES = 10
N_DIM = 10000

# Set up some utils
def sample_key_value(vsa: type[thd.VSATensor], n_dim: int, normalize: bool = True):
    key = vsa.random(num_vectors=1, dimensions=n_dim).flatten()
    value = vsa.random(num_vectors=1, dimensions=n_dim).flatten()
    if normalize:
        key = key.normalize()
        value = value.normalize()
    return {"key": key, "value": value}

def unbind(x: thd.VSATensor, y: thd.VSATensor) -> thd.VSATensor:
    return x.bind(y.inverse())

# Now, bind many pairs of VSATensors, then bundle them all together into a single VSATensor
vsa: type[thd.VSATensor] = thd.FHRRTensor

pairs = [
    sample_key_value(vsa, N_DIM, normalize=True)
    for _ in range(N_VALUES)
]
combined = thd.multiset(torch.stack([
    p["key"].bind(p["value"])
    for p in pairs
]))

print(
    unbind(combined, pairs[0]["key"]).cosine_similarity(pairs[0]["value"]) * np.sqrt(N_VALUES)  # Approx = 1
)
```

In other words, for the final cosine similarity to be ~1, I need to normalise by the sqrt of the number of bundled pairs. When using `HRRTensor`, this normalisation is unnecessary (though here I'm using `dot_similarity` not `cosine_similarity`):

```
vsa: type[thd.VSATensor] = thd.HRRTensor

pairs = [
    sample_key_value(vsa, N_DIM, normalize=True)
    for _ in range(N_VALUES)
]
combined = thd.multiset(torch.stack([
    p["key"].bind(p["value"])
    for p in pairs
]))

print(
    unbind(combined, pairs[0]["key"]).dot_similarity(pairs[0]["value"])  # Also approx = 1
)
```

Is this a property of the FHHR method, or am I missing something during the binding / bundling / unbinding operations? I've seen a similar need for normalisation of the final similarity score with `MAPTensor` (also `sqrt(N)`), `MCRTensor`, and `BSCTensor` (both `N`). 

I saw [this closed issue](https://github.com/hyperdimensional-computing/torchhd/issues/172) about normalisation, but it looks as though this resulted in adding the `normalize()` method which does not seem to change the result (at least, how I'm using it above).

The reason I ask is that my intended use case involves unbinding/querying against multiple different bound vectors (which may be formed of different numbers of bound pairs) and so this normalisation would be an additional value that needs tracking.

Thanks again! 🙏

**Comment by mikeheddes** [comment]
Hi @danmichaeljones, thanks for opening this issue. 

The normalization method behaves differently between FHRR and HRR. HRR uses real-valued vectors and `normalize` will set the L2 norm of the vector to be one. However, FHRR uses complex-valued vectors and `normalize` ensures that each element of the vector has a magnitude of one (i.e., falls on the complex unit circle). The normalized FHRR vector thus has an L2 norm of `sqrt(N_DIM)`. FHRR vectors are normalized this way because otherwise the binding operation wouldn't work as expected. Because of this difference the dot product between the same vector is 1 for HRR but `N_DIM` for FHRR.

I hope this explanation helps.

**Comment by danmichaeljones** [comment]
Hey @mikeheddes, thanks for the explanation - I think I'd confused myself by making `N_VALUES` / `N_DIM` too round numbers! As you explained, the below now gives me approx 1 with `thd.FHRRTensor` regardless of what `N_VALUES` / `N_DIM` are
```
unbind(combined, pairs[0]["key"]).dot_similarity(pairs[0]["value"]) / N_DIM  # Approx 1, as expected
```
Thanks for your help, I'll close this issue! 👍
